src/finding/summary.go

```diff
@@ -0,0 +1,139 @@
+package finding
+
+import (
+	"slices"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/ecr/types"
+	"github.com/cultureamp/ecrscanresults/findingconfig"
+)
+
+type Detail struct {
+	// The name associated with the finding, usually a CVE number.
+	Name string
+
+	URI string
+
+	// The description of the finding.
+	Description string
+
+	// The finding severity.
+	Severity types.FindingSeverity
+
+	PackageName    string
+	PackageVersion string
+	CVSS2Score     string
+	CVSS2Vector    string
+
+	Ignore *findingconfig.Ignore
+}
+
+type SeverityCount struct {
+	// Included is the number of findings that count towards the threshold for this severity.
+	Included int32
+
+	// Ignored is the number of findings that were ignored for the purposes of the threshold.
+	Ignored int32
+}
+
+type Summary struct {
+	// the counts by threshold, taking ignore configuration into account
+	Counts map[types.FindingSeverity]SeverityCount
+
+	Details []Detail
+
+	// the set of finding IDs that have been ignored by configuration
+	Ignored []Detail
+
+	// The time of the last completed image scan.
+	ImageScanCompletedAt *time.Time
+
+	// The time when the vulnerability data was last scanned.
+	VulnerabilitySourceUpdatedAt *time.Time
+}
+
+func (s *Summary) addDetail(d Detail) {
+	s.Details = append(s.Details, d)
+	s.updateCount(d.Severity, SeverityCount{Included: 1})
+}
+
+func (s *Summary) addIgnored(d Detail) {
+	s.Ignored = append(s.Ignored, d)
+	s.updateCount(d.Severity, SeverityCount{Ignored: 1})
+}
+
+func (s *Summary) updateCount(severity types.FindingSeverity, updateBy SeverityCount) {
+	counts := SeverityCount{}
+	if c, exists := s.Counts[severity]; exists {
+		counts = c
+	}
+
+	counts.Ignored += updateBy.Ignored
+	counts.Included += updateBy.Included
+
+	s.Counts[severity] = counts
+}
+
+func NewSummary() Summary {
```

Comment:
It just occurred to me that "New" may already [be a builtin](https://go.dev/src/builtin/builtin.go).

Reply:
I think this is a Go convention. It's a factory function and it's the equivalent of a constructor in Go

Reply:
It is a factory function as Cal says. The other thing is that I'm using a `_test` package for the tests so they can't use private APIs (another common practice) so I've exposed this on purpose. In some ways that's odd, but this isn't a library, so I'm less precious about API surface.

Reply:
> The other thing is that I'm using a _test package for the tests so they can't use private APIs (another common practice) so I've exposed this on purpose.

Like everything, this has pros and cons. Testing the code as though you're a package consumer should result in a better public interface, but then a lot more needs to be public so it can be tested. I think I'd prefer to have the tests in the same package, so they can directly access private functions and variables, and be disciplined about providing a good public interface

Reply:
This isn't used from the tests as I thought -- perhaps it was previously? I've made it private.

Reply:
That should be fine.

This just defines the shape required for `Summary` structs created by default when transforming the findings from AWS.